TestOnICEGatheringStateChange is flaky

```
=== RUN   TestOnICEGatheringStateChange
--- FAIL: TestOnICEGatheringStateChange (10.00s)
    peerconnection_go_test.go:668: Completed before gathering
    peerconnection_go_test.go:686: Gathering and Complete were never seen
=== RUN   TestPeerConnectionTrickle
goroutine profile: total 5
1 @ 0x805f697 0x8052413 0x808801f 0x83ee687 0x83eeb27 0x83ee5ed 0x8419d50 0x8110c2a 0x809e321
#	0x83ee686	github.com/pion/transport/test.getRoutines+0x36			/go/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:51
#	0x83eeb26	github.com/pion/transport/test.CheckRoutines.func1+0xb6		/go/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:32
#	0x83ee5ec	github.com/pion/transport/test.CheckRoutines+0x5c		/go/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:44
#	0x8419d4f	github.com/pion/webrtc/v2.TestPeerConnectionTrickle+0x7f	/go/src/github.com/pion/webrtc/peerconnection_go_test.go:706
#	0x8110c29	testing.tRunner+0x99						/usr/local/go/src/testing/testing.go:909

1 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x8110f5b 0x8113daa 0x8110c2a 0x8111fcd 0x811127b 0x844de44 0x80754d2 0x809e321
#	0x8110f5a	testing.(*T).Run+0x2ca		/usr/local/go/src/testing/testing.go:961
#	0x8113da9	testing.runTests.func1+0x59	/usr/local/go/src/testing/testing.go:1202
#	0x8110c29	testing.tRunner+0x99		/usr/local/go/src/testing/testing.go:909
#	0x8111fcc	testing.runTests+0x22c		/usr/local/go/src/testing/testing.go:1200
#	0x811127a	testing.(*M).Run+0x13a		/usr/local/go/src/testing/testing.go:1117
#	0x844de43	main.main+0x103			_testmain.go:322
#	0x80754d1	runtime.main+0x201		/usr/local/go/src/runtime/proc.go:203

1 @ 0x80758a4 0x8070524 0x806f9fb 0x80be807 0x80bf396 0x80bf37b 0x81bb7ef 0x81cf63f 0x81ce407 0x82ede10 0x809e321
#	0x806f9fa	internal/poll.runtime_pollWait+0x4a			/usr/local/go/src/runtime/netpoll.go:184
#	0x80be806	internal/poll.(*pollDesc).wait+0x36			/usr/local/go/src/internal/poll/fd_poll_runtime.go:87
#	0x80bf395	internal/poll.(*pollDesc).waitRead+0x165		/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
#	0x80bf37a	internal/poll.(*FD).ReadFrom+0x14a			/usr/local/go/src/internal/poll/fd_unix.go:219
#	0x81bb7ee	net.(*netFD).readFrom+0x3e				/usr/local/go/src/net/fd_unix.go:208
#	0x81cf63e	net.(*UDPConn).readFrom+0x3e				/usr/local/go/src/net/udpsock_posix.go:47
#	0x81ce406	net.(*UDPConn).ReadFrom+0x46				/usr/local/go/src/net/udpsock.go:121
#	0x82ede0f	github.com/pion/ice.(*candidateBase).recvLoop+0xef	/go/pkg/mod/github.com/pion/ice@v0.7.10/candidate_base.go:92

1 @ 0x80758a4 0x8070524 0x806f9fb 0x80be807 0x80bff9d 0x80bff7e 0x81c9cb8 0x82c7c2c 0x82cb232 0x82cb211 0x82cee21 0x809e321
#	0x806f9fa	internal/poll.runtime_pollWait+0x4a			/usr/local/go/src/runtime/netpoll.go:184
#	0x80be806	internal/poll.(*pollDesc).wait+0x36			/usr/local/go/src/internal/poll/fd_poll_runtime.go:87
#	0x80bff9c	internal/poll.(*pollDesc).waitRead+0xec			/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
#	0x80bff7d	internal/poll.(*FD).RawRead+0xcd			/usr/local/go/src/internal/poll/fd_unix.go:534
#	0x81c9cb7	net.(*rawConn).Read+0x47				/usr/local/go/src/net/rawconn.go:43
#	0x82c7c2b	golang.org/x/net/internal/socket.(*Conn).recvMsg+0x18b	/go/pkg/mod/golang.org/x/net@v0.0.0-20200301022130-244492dfa37a/internal/socket/rawconn_msg.go:32
#	0x82cb231	golang.org/x/net/internal/socket.(*Conn).RecvMsg+0x141	/go/pkg/mod/golang.org/x/net@v0.0.0-20200301022130-244492dfa37a/internal/socket/socket.go:255
#	0x82cb210	golang.org/x/net/ipv4.(*payloadHandler).ReadFrom+0x120	/go/pkg/mod/golang.org/x/net@v0.0.0-20200301022130-244492dfa37a/ipv4/payload_cmsg.go:31
#	0x82cee20	github.com/pion/mdns.(*Conn).start+0xf0			/go/pkg/mod/github.com/pion/mdns@v0.0.4/conn.go:256

1 @ 0x814626c 0x81460da 0x8143133 0x83ee9d7 0x809e321
#	0x814626b	runtime/pprof.writeRuntimeProfile+0x7b			/usr/local/go/src/runtime/pprof/pprof.go:708
#	0x81460d9	runtime/pprof.writeGoroutine+0x79			/usr/local/go/src/runtime/pprof/pprof.go:670
#	0x8143132	runtime/pprof.(*Profile).WriteTo+0x2c2			/usr/local/go/src/runtime/pprof/pprof.go:329
#	0x83ee9d6	github.com/pion/transport/test.TimeOut.func1+0x56	/go/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:18

panic: timeout
```

--- a/icegatherer.go
+++ b/icegatherer.go
@@ -4,6 +4,7 @@ package webrtc
 
 import (
 	"sync"
+	"sync/atomic"
 
 	"github.com/pion/ice"
 	"github.com/pion/logging"
@@ -23,8 +24,8 @@ type ICEGatherer struct {
 
 	agent *ice.Agent
 
-	onLocalCandidateHdlr func(candidate *ICECandidate)
-	onStateChangeHdlr    func(state ICEGathererState)
+	onLocalCandidateHdlr atomic.Value // func(candidate *ICECandidate)
+	onStateChangeHdlr    atomic.Value // func(state ICEGathererState)
 
 	api *API
 }
@@ -124,7 +125,7 @@ func (g *ICEGatherer) createAgent() error {
 
 	g.agent = agent
 	if !g.api.settingEngine.candidates.ICETrickle {
-		g.state = ICEGathererStateComplete
+		atomicStoreICEGathererState(&g.state, ICEGathererStateComplete)
 	}
 
 	return nil
@@ -136,16 +137,15 @@ func (g *ICEGatherer) Gather() error {
 		return err
 	}
 
-	g.lock.Lock()
-	defer g.lock.Unlock()
-
-	onLocalCandidateHdlr := g.onLocalCandidateHdlr
-	if onLocalCandidateHdlr == nil {
-		onLocalCandidateHdlr = func(*ICECandidate) {}
+	onLocalCandidateHdlr := func(*ICECandidate) {}
+	if hdlr, ok := g.onLocalCandidateHdlr.Load().(func(candidate *ICECandidate)); ok && hdlr != nil {
+		onLocalCandidateHdlr = hdlr
 	}
 
+	g.lock.Lock()
 	isTrickle := g.api.settingEngine.candidates.ICETrickle
 	agent := g.agent
+	g.lock.Unlock()
 
 	if !isTrickle {
 		return nil
@@ -161,9 +161,7 @@ func (g *ICEGatherer) Gather() error {
 			}
 			onLocalCandidateHdlr(&c)
 		} else {
-			g.lock.Lock()
 			g.setState(ICEGathererStateComplete)
-			g.lock.Unlock()
 
 			onLocalCandidateHdlr(nil)
 		}
@@ -220,29 +218,24 @@ func (g *ICEGatherer) GetLocalCandidates() ([]ICECandidate, error) {
 // OnLocalCandidate sets an event handler which fires when a new local ICE candidate is available
 // Take note that the handler is gonna be called with a nil pointer when gathering is finished.
 func (g *ICEGatherer) OnLocalCandidate(f func(*ICECandidate)) {
-	g.lock.Lock()
-	defer g.lock.Unlock()
-	g.onLocalCandidateHdlr = f
+	g.onLocalCandidateHdlr.Store(f)
 }
 
 // OnStateChange fires any time the ICEGatherer changes
 func (g *ICEGatherer) OnStateChange(f func(ICEGathererState)) {
-	g.lock.Lock()
-	defer g.lock.Unlock()
-	g.onStateChangeHdlr = f
+	g.onStateChangeHdlr.Store(f)
 }
 
 // State indicates the current state of the ICE gatherer.
 func (g *ICEGatherer) State() ICEGathererState {
-	g.lock.RLock()
-	defer g.lock.RUnlock()
-	return g.state
+	return atomicLoadICEGathererState(&g.state)
 }
 
 func (g *ICEGatherer) setState(s ICEGathererState) {
-	g.state = s
-	if hdlr := g.onStateChangeHdlr; hdlr != nil {
-		go hdlr(s)
+	atomicStoreICEGathererState(&g.state, s)
+
+	if hdlr, ok := g.onStateChangeHdlr.Load().(func(state ICEGathererState)); ok && hdlr != nil {
+		hdlr(s)
 	}
 }
 
@@ -260,9 +253,10 @@ func (g *ICEGatherer) SignalCandidates() error {
 		return err
 	}
 
-	g.lock.Lock()
-	onLocalCandidateHdlr := g.onLocalCandidateHdlr
-	g.lock.Unlock()
+	var onLocalCandidateHdlr func(*ICECandidate)
+	if hdlr, ok := g.onLocalCandidateHdlr.Load().(func(candidate *ICECandidate)); ok {
+		onLocalCandidateHdlr = hdlr
+	}
 
 	if onLocalCandidateHdlr != nil {
 		go func() {

--- a/icegathererstate.go
+++ b/icegathererstate.go
@@ -1,7 +1,11 @@
 package webrtc
 
+import (
+	"sync/atomic"
+)
+
 // ICEGathererState represents the current state of the ICE gatherer.
-type ICEGathererState byte
+type ICEGathererState uint32
 
 const (
 	// ICEGathererStateNew indicates object has been created but
@@ -33,4 +37,12 @@ func (s ICEGathererState) String() string {
 	default:
 		return unknownStr
 	}
+}
+
+func atomicStoreICEGathererState(state *ICEGathererState, newState ICEGathererState) {
+	atomic.StoreUint32((*uint32)(state), uint32(newState))
+}
+
+func atomicLoadICEGathererState(state *ICEGathererState) ICEGathererState {
+	return ICEGathererState(atomic.LoadUint32((*uint32)(state)))
 }

--- a/peerconnection_go_test.go
+++ b/peerconnection_go_test.go
@@ -658,14 +658,23 @@ func TestOnICEGatheringStateChange(t *testing.T) {
 	peerConn, err := NewAPI(WithSettingEngine(s)).NewPeerConnection(Configuration{})
 	assert.NoError(t, err)
 
-	peerConn.OnICEGatheringStateChange(func(s ICEGathererState) {
+	var onStateChange func(s ICEGathererState)
+	onStateChange = func(s ICEGathererState) {
+		// Access to ICEGatherer in the callback must not cause dead lock.
+		peerConn.OnICEGatheringStateChange(onStateChange)
+		if state := peerConn.iceGatherer.State(); state != s {
+			t.Errorf("State change callback argument (%s) and State() (%s) result differs",
+				s, state,
+			)
+		}
+
 		switch s {
 		case ICEGathererStateClosed:
 			close(seenClosed)
 			return
 		case ICEGathererStateGathering:
 			if seenComplete.get() {
-				t.Fatal("Completed before gathering")
+				t.Error("Completed before gathering")
 			}
 			seenGathering.set(true)
 		case ICEGathererStateComplete:
@@ -675,7 +684,8 @@ func TestOnICEGatheringStateChange(t *testing.T) {
 		if seenGathering.get() && seenComplete.get() {
 			close(seenGatheringAndComplete)
 		}
-	})
+	}
+	peerConn.OnICEGatheringStateChange(onStateChange)
 
 	offer, err := peerConn.CreateOffer(nil)
 	assert.NoError(t, err)

